returnn/tf/layers/basic.py

```diff
@@ -913,44 +913,227 @@ def transform_config_dict(cls, d, network, get_layer):
+    else:
+      return super(GatherLayer, self).get_dep_layers()
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, sources, position, axis, **kwargs):
     """
     :param str name:
     :param list[LayerBase] sources:
+    :param LayerBase|int position:
     :param str axis:
     :rtype: Data
     """
-    out = get_concat_sources_data_template(sources, name="%s_output" % name)
-    axis = out.get_axis_from_description(axis, allow_int=False)
-    return out.copy_template_excluding_axis(axis)
+    input_data = get_concat_sources_data_template(sources)
+    old_gather_axis = input_data.get_axis_from_description(axis, allow_int=False)
+
+    if isinstance(position, int):
+      position_data = Data(name="constant_position", dtype="int32", shape=(), batch_dim_axis=None)
+    else:
+      position_data = position.output.copy_template()
+    assert position_data.dtype in ["int32", "int64"]
+
+    # determine all common axes of input_data and position_data
+    common_axes_input, common_axes_position, input_axes, position_axes = (
+      cls._get_common_input_position_axes(input_data, position_data, old_gather_axis))
+    batch_dims = len(common_axes_input)
+
+    # move gather axis back if necessary s.t. common axes are always in front
+    gather_axis = max(batch_dims, old_gather_axis)
+
+    # (BatchAxes.., InputAxesBeforeGatherAxis, PositionAxes.., InputAxesAfterGatherAxis..)
+    shape = (
+      [input_data.batch_shape[ax] for ax in common_axes_input if ax != input_data.batch_dim_axis] +
+      [input_data.batch_shape[ax] for ax in input_axes[:gather_axis-batch_dims] if ax != input_data.batch_dim_axis] +
+      [position_data.batch_shape[ax] for ax in position_axes if ax != position_data.batch_dim_axis] +
+      [input_data.batch_shape[ax] for ax in input_axes[gather_axis-batch_dims:] if ax != input_data.batch_dim_axis])
+    out_type = input_data.get_kwargs()
+    out_type["name"] = "%s_output" % name
+    out_type["shape"] = shape
+    out_type["beam"] = SearchBeam.get_combined_beam(input_data.beam, position_data.beam)
+    out_type["available_for_inference"] = input_data.available_for_inference and position_data.available_for_inference
+
+    # Take axes from input_data if they exist there, otherwise from position_data
+    for axis_kind in Data.SpecialAxesNames:
+      input_axis, position_axis = getattr(input_data, axis_kind), getattr(position_data, axis_kind)
+      if input_axis is not None and input_axis != old_gather_axis:
+        out_type[axis_kind] = cls._translate_input_axis(
+          input_axis, old_gather_axis, common_axes_input, input_axes, position_axes)
+      elif position_axis is not None:
+        out_type[axis_kind] = cls._translate_position_axis(
+          position_axis, old_gather_axis, common_axes_position, position_axes)
+      else:
+        out_type[axis_kind] = None
+    if input_data.feature_dim_axis_or_unspecified is NotSpecified:
+      out_type["feature_dim_axis"] = NotSpecified
```

Comment:
Maybe the code should be restructured and simplified. Or this case should just be kept separate. Not sure how to simplify it in a clean way. But the logic should be like:
```suggestion
    if input_data.feature_dim_axis_or_unspecified is NotSpecified and position_data.feature_dim_axis_or_unspecified is NotSpecified:
      out_type["feature_dim_axis"] = NotSpecified
    elif input_data.feature_dim_axis_or_unspecified is NotSpecified:
      assert position_data.feature_dim_axis_or_unspecified is not NotSpecified
      if position_data.feature_dim_axis is not None:
        out_type["feature_dim_axis"] = cls._translate_position_axis(
          position_data.feature_dim_axis, old_gather_axis, common_axes_position, position_axes)
      else:
        out_type["feature_dim_axis"] = NotSpecified
    elif position_data.feature_dim_axis_or_unspecified is NotSpecified:
      assert input_data.feature_dim_axis_or_unspecified is not NotSpecified
      if input_data.feature_dim_axis is not None:
        out_type["feature_dim_axis"] = cls._translate_position_axis(
          input_data.feature_dim_axis, old_gather_axis, common_axes_position, input_axes)
      else:
        out_type["feature_dim_axis"] = NotSpecified
    else:
      assert input_data.feature_dim_axis_or_unspecified is not NotSpecified
      assert position_data.feature_dim_axis_or_unspecified is not NotSpecified
      pass  # keep the logic as before
```

Reply:
What a beast! :D For now I just used your suggestion (there was a typo in line 1120, it should be `cls._translate_input_axis(input_data.feature_dim_axis, ...)`), we may change that later.